crates/goose/src/model.rs

```diff
@@ -1,323 +1,346 @@
+    #[error("Value for '{0}' is out of valid range: {1}")]
+    InvalidRange(String, String),
+}
+
+static MODEL_SPECIFIC_LIMITS: Lazy<Vec<(&'static str, usize)>> = Lazy::new(|| {
+    vec![
+        // openai
+        ("gpt-4-turbo", 128_000),
+        ("gpt-4.1", 1_000_000),
+        ("gpt-4-1", 1_000_000),
+        ("gpt-4o", 128_000),
+        ("o4-mini", 200_000),
+        ("o3-mini", 200_000),
+        ("o3", 200_000),
+        // anthropic - all 200k
+        ("claude", 200_000),
+        // google
+        ("gemini-1", 128_000),
+        ("gemini-2", 1_000_000),
+        ("gemma-3-27b", 128_000),
+        ("gemma-3-12b", 128_000),
+        ("gemma-3-4b", 128_000),
+        ("gemma-3-1b", 32_000),
+        ("gemma3-27b", 128_000),
+        ("gemma3-12b", 128_000),
+        ("gemma3-4b", 128_000),
+        ("gemma3-1b", 32_000),
+        ("gemma-2-27b", 8_192),
+        ("gemma-2-9b", 8_192),
+        ("gemma-2-2b", 8_192),
+        ("gemma2-", 8_192),
+        ("gemma-7b", 8_192),
+        ("gemma-2b", 8_192),
+        ("gemma1", 8_192),
+        ("gemma", 8_192),
+        // facebook
+        ("llama-2-1b", 32_000),
+        ("llama", 128_000),
+        // qwen
+        ("qwen3-coder", 262_144),
+        ("qwen2-7b", 128_000),
+        ("qwen2-14b", 128_000),
+        ("qwen2-32b", 131_072),
+        ("qwen2-70b", 262_144),
+        ("qwen2", 128_000),
+        ("qwen3-32b", 131_072),
+        // other
+        ("kimi-k2", 131_072),
+        ("grok-4", 256_000),
+        ("grok", 131_072),
+    ]
 });
 
-/// Configuration for model-specific settings and limits
 #[derive(Debug, Clone, Serialize, Deserialize)]
 pub struct ModelConfig {
-    /// The name of the model to use
     pub model_name: String,
-    /// Optional explicit context limit that overrides any defaults
     pub context_limit: Option<usize>,
-    /// Optional temperature setting (0.0 - 1.0)
     pub temperature: Option<f32>,
-    /// Optional maximum tokens to generate
     pub max_tokens: Option<i32>,
-    /// Whether to interpret tool calls with toolshim
     pub toolshim: bool,
-    /// Model to use for toolshim (optional as a default exists)
     pub toolshim_model: Option<String>,
 }
 
-/// Struct to represent model pattern matches and their limits
 #[derive(Debug, Clone, Serialize, Deserialize)]
 pub struct ModelLimitConfig {
     pub pattern: String,
     pub context_limit: usize,
 }
 
 impl ModelConfig {
-    /// Create a new ModelConfig with the specified model name
-    ///
-    /// The context limit is set with the following precedence:
-    /// 1. Explicit context_limit if provided in config
-    /// 2. Environment variable override (GOOSE_CONTEXT_LIMIT)
-    /// 3. Model-specific default based on model name
-    /// 4. Global default (128_000) (in get_context_limit)
-    pub fn new(model_name: String) -> Self {
-        Self::new_with_context_env(model_name, None)
+    pub fn new(model_name: &str) -> Result<Self, ConfigError> {
```

Comment:
@DOsinga this is the very wide reaching change, but it does seem nicely idiomatic rust: (eg https://doc.rust-lang.org/std/fs/struct.File.html#method.open canonical example). Lots of clients to this use `?` which I assume is ok when we just want to let it fail?